File: Bots/Phase2Bot.js
//IMPORTS
const config = require("./config")
const fs = require('fs')
const filebuffer = fs.readFileSync('blackjack.db')

const {randomBytes} = require("crypto")

//GLOBAL VARS
const Web3 = require("web3");
const randomBinary = require("random-binary")
const sqlite3 = require("sqlite3").verbose()
const {open} = require('sqlite');

const initSqlJs = require('sql.js')
const provider = new Web3.providers.HttpProvider(config.web3Provider);
const web3 = new Web3(provider);


web3.eth.net.isListening()
.then(() => console.log('web3 is connected'))
.catch(e => console.log('Wow. Something went wrong'));

const casinoContract = new web3.eth.Contract(config.abi,config.smartContractAddress)


web3.eth.accounts.wallet.add('0x' + config.casinoPrivateKey )


const accounts = web3.eth.accounts.wallet
const account = accounts["0"]
const myAddress = account["address"]

function createDBConnection(filename){
     return open({
          filename,
          driver: sqlite3.Database
     })
}





const generateRandomNumberWithHash=()=>{
       let rCom1 = randomBinary(156).toString()
       let rCom2 = randomBinary(156).toString()

       rCom1HashBytes32 = Web3.utils.soliditySha3({type: 'string', value: rCom1})
       rCom2HashBytes32 = Web3.utils.soliditySha3({type: 'string', value: rCom2})

       return{
            "rCom1": rCom1,
            "rCom1Hash": rCom1HashBytes32,
            "rCom2": rCom2,
            "rCom2Hash": rCom2HashBytes32
       }
}


//GOOD?
const shuffleCards=(shuffleArray)=>{
       let deck = []



       for(var i=0; i<52;i++){
              deck.push(i)
       }

       var currentIndex = 52
       var count= 0
       var temporaryValue = 0

       while(currentIndex != 0){
           currentIndex = currentIndex - 1
            var randomIndex = shuffleArray[count]
            temporaryValue = deck[currentIndex]
            deck[currentIndex] = deck[randomIndex]
            deck[randomIndex] = temporaryValue
            count += 1

       }
       return deck
}


//GOOD?
const getPhase2Games=async ()=>{
       const temp = await casinoContract.methods.getPhase2Games().call({from: myAddress})
       
       activeGames = []
       for (i = 0; i < temp.length; i++){

              if (temp[i] != '0x0000000000000000000000000000000000000000'){
                     activeGames.push(temp[i])
              }
       }

       return activeGames
}
// const test=()=>{

//      db.serialize(()=>{
//           db.all("SELECT * FROM GAMEDATA WHERE [ADR] = ?", '0x2372F830e274e0f4B5CD7710d519B60eA06007CB', (error, allRows)=>{
                 
//           }
//      }
// }

const getGameData =async(activeGame)=>{

     db = await createDBConnection("./blackjack.db")

     const query = "SELECT * FROM GAMEDATA WHERE [ADR] = ?"
     const row = await db.get(query,[activeGame])
     
     return row

}

const phase2Response=async (activeGame, nonce)=>{

       
       try{
              console.log("Processing game for: ", activeGame)
              
              var gameData = await getGameData(activeGame)

              console.log("GAME DATA",gameData)

              if (gameData == null){
                   casinoData = await generateRandomNumberWithHash()

                   const result = await db.run("INSERT OR REPLACE INTO GAMEDATA(ADR,RCOM1,RCOM2,RCOM1HASH,RCOM2HASH,RP1,RP2) VALUES (:ADR,:RCOM1,:RCOM2,:RCOM1HASH,:RCOM2HASH,:RP1,:RP2)",{
                        ":ADR": activeGame,
                        ":RCOM1": casinoData["rCom1"],
                        ":RCOM2": casinoData["rCom2"],
                        ":RCOM1HASH": casinoData["rCom1Hash"],
                        ":RCOM2HASH": casinoData["rCom2Hash"],
                        ":RP1": "",
                        ":RP2": ""
                   })
                   //Do db insert here

                   console.log("Created rcom1,rcom2,hash1,hash2 for casino. Waiting for player rp1 and rp2.")
              }
              else if ((gameData["RP1"].length == 0) || (gameData["RP2"] == 0)){
                        console.log("Cannot create deck yet. Still waiting on player rp1 and rp2");
               }
               else{
                        console.log("All data is present! Will now work on creating deck...")


                        var rCom1 = gameData["RCOM1"]
                        var rCom2 = gameData["RCOM2"]
                        var rP1 = gameData["RP1"]
                        var rP2 = gameData["RP2"]
                        var rCom1Hash = gameData["RCOM1HASH"]
                        var rCom2Hash = gameData["RCOM2HASH"]
                        var rFyBuild = []

                        var rD = rCom1 + rCom2
                        var rP = rP1 + rP2


                        for(var i=0; i<312; i++){
                               if (rD[i] === rP[i]){
                                      rFyBuild.push("0")
                               }else{
                                      rFyBuild.push("1")
                               }
                        }
                        rFy = rFyBuild.join('')
                        let binaryList = rFy.match(/.{6}/g)

                        let shuffleList = BinaryToDec(binaryList)

                        let shuffledDeck = shuffleCards(shuffleList)

                    await callTransaction(activeGame,shuffledDeck,gameData["RCOM1HASH"],gameData["RCOM2HASH"], nonce)
                        
                   
              }

       }catch(error){
              console.log(error)
       }
}

const BinaryToDec =(array)=>{
     let newArr = []
     for (var i=0;i<array.length;i++){

            if(parseInt(array[i], 2) >= 52){
                   newArr.push((parseInt(array[i], 2)) % 52)
            }else{
                   newArr.push(parseInt(array[i], 2))
            }
     }
     return newArr
}

const callTransaction= async (activeGame,shuffledDeck,rCom1Hash,rCom2Hash, nonce)=>{
     console.log(activeGame)
     console.log(shuffledDeck)
     console.log(typeof rCom1Hash)
     console.log(rCom2Hash)
     const gasPrice = await web3.eth.getGasPrice();

     const gasEstimate = await casinoContract.methods.Casino_get_deck(activeGame,shuffledDeck,rCom1Hash,rCom2Hash).estimateGas({ from: myAddress, to: activeGame  });

     const response = await casinoContract.methods.Casino_get_deck(activeGame,shuffledDeck,rCom1Hash,rCom2Hash).send({from: myAddress, to: activeGame,gasPrice: gasPrice, gas: gasEstimate })

     console.log("Done!")
     console.log(response)
     console.log()

}


const main = async () => {
   
     nonce = await web3.eth.getTransactionCount(myAddress, 'pending')

     const phase2Bot =  async () => {
          console.log("- - - - - - - - - -")
          console.log("Scanning games...")
          console.log()
   
          var phase2Games = await getPhase2Games()
          console.log(phase2Games)

          let randInt = Math.floor(Math.random() * 1000)
          if((phase2Games.length)>0){
                 console.log("Games in phase 2 detected:")
   
                 for(i=0;i<phase2Games.length;i++){
                    console.log(phase2Games[i])
                    nonce += randInt
                    await phase2Response(phase2Games[i], nonce)
                 }
   
          }
          setTimeout(phase2Bot, 5000);
     }
     phase2Bot()
     
 }
 
main()



      
   
   

File: Bots/Phase2Bot.py

from web3 import Web3
import config
import random
import time
import asyncio
import sqlite3
from Crypto.Random import random

def generateRandomNumberWithHash():
       rComNum = random.getrandbits(512) #int
       rCom = bin(rComNum).split('b')[1] #string
       rComHash = str(Web3.keccak(text=rCom))

       hashBuild = []
       for index,char in enumerate(rComHash):
              if index == 0 or index == 1 or index == (len(rComHash) - 1):
                     continue
              else:
                     hashBuild.append(char)
       
       rComHash = "".join(hashBuild)
       rDBinaryList = []
       rd = ""
       for index,char in enumerate(rCom):
              if index == 312:
                     rd = "".join(rDBinaryList)
                     break
              else:
                     rDBinaryList.append(char)
       return {
              "rCom": rCom,
              "hash": rComHash,
              "rD": rd
       }
                     


w3 = Web3(Web3.HTTPProvider(config.infuraProvider))
accounts = w3.eth.account.privateKeyToAccount(config.casinoPrivateKey)
account = accounts.address
smartContractAddress= Web3.toChecksumAddress(config.smartContractAddress)
casinoContract = w3.eth.contract(address= smartContractAddress, abi= config.abi)
connection = sqlite3.connect(r"./blackjack.db")
cursor = connection.cursor()

def shuffleCards(shuffleArray):

    deck = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51]

    for index,val in enumerate(deck):
        deck[index],deck[shuffleArray[index]] = deck[shuffleArray[index]],deck[index]
    
    return deck

def getPhase2Games():
       temp = casinoContract.functions.getPhase2Games().call({'from': account})
       activeGames= []
       for x in temp:
              if x != '0x0000000000000000000000000000000000000000':
                     activeGames.append(x)
       
       return activeGames

async def asyncFunction(delay,func):
       await asyncio.sleep(delay)
       return func

async def phase2Response(activeGame,nonce):

       try:
              print("Processing game for {}:".format(str(activeGame)))

              cursor.execute("SELECT * FROM GAMEDATA WHERE [ADR] = ?",(activeGame,))
              
              gameData = cursor.fetchall()
              
              if len(gameData) > 1:
                     print("Error: More than one entry of data")
              elif len(gameData) == 0:
                     casinoData = generateRandomNumberWithHash()
                     cursor.execute("INSERT OR REPLACE INTO GAMEDATA VALUES (?,?,?,?,?)",tuple([activeGame,casinoData["rCom"],casinoData["rD"],casinoData["hash"],""]))
                     connection.commit()
                     print("\tCreated rcom, rd, and hash for casino. Waiting for player rp.")
              else:
                     gameData = gameData[0]
                     if len(gameData[4]) == 0:
                            print("\t Cannot create deck yet. Still waiting on player rp.")
                     else:
                            print("\t All data is present! Will now work on creating deck.")
                            cursor.execute("SELECT * FROM GAMEDATA WHERE [ADR] = ?",(activeGame,))
                            gameData = cursor.fetchall()
                            gameData = gameData[0]

                            rD = gameData[2]
                            rP = gameData[4]
                            rComHash = gameData[3]
                            rFyBuild = []

                            for i in range(312):
                                   if rD[i] == rP[i]:
                                          rFyBuild.append("0")
                                   else:
                                          rFyBuild.append("1")
                            rFy = "".join(rFyBuild)

                            binaryList = [rFy[i:i+6] for i in range(0, len(rFy), 6)]

                            shuffleList = []
                            for binaryNumber in binaryList:
                                   num = int(binaryNumber,2)
                                   shuffleList.append(num%52)
                            


                            shuffledDeck = shuffleCards(shuffleList)

                            transaction = await asyncFunction(1,casinoContract.functions.Casino_get_deck(activeGame,shuffledDeck,rComHash).buildTransaction({
                                   'from': account,
                                   'nonce': nonce
                            }))

                            signed_tx = await asyncFunction(1,w3.eth.account.sign_transaction(transaction, config.casinoPrivateKey))

                            response = await asyncFunction(1,w3.eth.send_raw_transaction(signed_tx.rawTransaction))
                            print("\tDone:",response)
                            nonce += 1
       except Exception as e:
              print(str(e))


nonce = w3.eth.getTransactionCount(account)
while True:
       print("- - - - - - - - - -")
       print("Scanning games...")
       print()
       phase2Games = getPhase2Games()
       print(phase2Games)
       if len(phase2Games) > 0:
              print("Games in phase 2 detected:")
              for game in phase2Games:
                     print("\t{}".format(str(game)))
              print()
              for game in phase2Games:
                     asyncio.run(phase2Response(game,nonce))
                     nonce += 1

       time.sleep(5)